cpp/src/arrow/io/compressed.cc

```diff
@@ -330,19 +333,24 @@ class CompressedInputStream::Impl {
 
       // At this point, no more decompressed data remains,
       // so we need to decompress more
-      if (decompressor_->IsFinished()) {
-        break;
-      }
       // First try to read data from the decompressor
       if (compressed_) {
+        if (decompressor_->IsFinished()) {
```

Comment:
I spent a good 30 minutes trying to figure out the overall control flow (I fell asleep in bed!) of this loop. Maybe this warrant a refactoring/simplification? I won't mark this as blocking, but just a note that I can't comment if the change is sound.

Reply:
Well, I don't think this can be simplified. Is there a particular pain point? And/or should I add some comments (where?)?

Reply:
The following patch divides the function in 2 sub-functions, which I think, makes this a lot more readable (the book keeping of properties is hidden in each method).

```
diff --git a/cpp/src/arrow/io/compressed.cc b/cpp/src/arrow/io/compressed.cc
index 10eaf6380..f5148172d 100644
--- a/cpp/src/arrow/io/compressed.cc
+++ b/cpp/src/arrow/io/compressed.cc
@@ -308,53 +308,24 @@ class CompressedInputStream::Impl {
   Status Read(int64_t nbytes, int64_t* bytes_read, void* out) {
     std::lock_guard<std::mutex> guard(lock_);
 
-    *bytes_read = 0;
     auto out_data = reinterpret_cast<uint8_t*>(out);
 
-    while (nbytes > 0) {
-      int64_t avail = decompressed_ ? (decompressed_->size() - decompressed_pos_) : 0;
-      if (avail > 0) {
-        // Pending decompressed data is available, use it
-        avail = std::min(avail, nbytes);
-        memcpy(out_data, decompressed_->data() + decompressed_pos_, avail);
-        decompressed_pos_ += avail;
-        out_data += avail;
-        *bytes_read += avail;
-        nbytes -= avail;
-        if (decompressed_pos_ == decompressed_->size()) {
-          // Decompressed data is exhausted, release buffer
-          decompressed_.reset();
-        }
-        if (nbytes == 0) {
-          // We're done
-          break;
-        }
-      }
+    int64_t total_read = 0;
+    bool decompressor_has_data = true;
 
-      // At this point, no more decompressed data remains,
-      // so we need to decompress more
-      // First try to read data from the decompressor
-      if (compressed_) {
-        if (decompressor_->IsFinished()) {
-          // We just went over the end of a previous compressed stream.
-          RETURN_NOT_OK(decompressor_->Reset());
-          fresh_decompressor_ = true;
-        }
-        RETURN_NOT_OK(DecompressData());
-      }
-      if (!decompressed_ || decompressed_->size() == 0) {
-        // Got nothing, need to read more compressed data
-        RETURN_NOT_OK(EnsureCompressedData());
-        if (compressed_pos_ == compressed_->size()) {
-          // No more data to decompress
-          if (!fresh_decompressor_) {
-            return Status::IOError("Truncated compressed stream");
-          }
-          break;
-        }
-        RETURN_NOT_OK(DecompressData());
+    while (nbytes - total_read > 0 && decompressor_has_data) {
+      total_read += ReadFromDecompressed(nbytes - total_read, out_data + total_read);
+
+      if (nbytes == total_read) {
+        break;
       }
+
+      // At this point, no more decompressed data remains, so we need to
+      // decompress more
+      RETURN_NOT_OK(WriteToDecompressed(&decompressor_has_data));
     }
+
+    *bytes_read = total_read;
     return Status::OK();
   }
 
@@ -371,6 +342,50 @@ class CompressedInputStream::Impl {
   std::shared_ptr<InputStream> raw() const { return raw_; }
 
  private:
+  int64_t ReadFromDecompressed(int64_t nbytes, uint8_t* out) {
+    int64_t readable = decompressed_ ? (decompressed_->size() - decompressed_pos_) : 0;
+    int64_t read_bytes = std::min(readable, nbytes);
+
+    if (read_bytes > 0) {
+      memcpy(out, decompressed_->data() + decompressed_pos_, read_bytes);
+      decompressed_pos_ += read_bytes;
+
+      if (decompressed_pos_ == decompressed_->size()) {
+        // Decompressed data is exhausted, release buffer
+        decompressed_.reset();
+      }
+    }
+
+    return read_bytes;
+  }
+
+  Status WriteToDecompressed(bool* has_data) {
+    // First try to read data from the decompressor
+    if (compressed_) {
+      if (decompressor_->IsFinished()) {
+        // We just went over the end of a previous compressed stream.
+        RETURN_NOT_OK(decompressor_->Reset());
+        fresh_decompressor_ = true;
+      }
+      RETURN_NOT_OK(DecompressData());
+    }
+    if (!decompressed_ || decompressed_->size() == 0) {
+      // Got nothing, need to read more compressed data
+      RETURN_NOT_OK(EnsureCompressedData());
+      if (compressed_pos_ == compressed_->size()) {
+        // No more data to decompress
+        if (!fresh_decompressor_) {
+          return Status::IOError("Truncated compressed stream");
+        }
+        *has_data = false;
+        return Status::OK();
+      }
+      RETURN_NOT_OK(DecompressData());
+    }
+
+    return Status::OK();
+  }
+
   // Read 64 KB compressed data at a time
   static const int64_t kChunkSize = 64 * 1024;
   // Decompress 1 MB at a time
```

Reply:
Thanks! I'll apply it.